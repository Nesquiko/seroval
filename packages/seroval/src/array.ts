export function forEach<T>(arr: T[], callback: (value: T, index: number) => (boolean | void)) {
  for (let i = 0, len = arr.length; i < len; i++) {
    if (callback(arr[i], i)) {
      break;
    }
  }
}

export function join(arr: string[], joiner: string): string {
  let result = '';
  forEach(arr, (value, index) => {
    result += value;
    if (index < arr.length - 1) {
      result += joiner;
    }
  });
  return result;
}
