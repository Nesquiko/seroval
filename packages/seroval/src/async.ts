/* eslint-disable no-await-in-loop */
import {
  SerializationContext,
  insertRef,
  constructorCheck,
  getRefParam,
  createRef,
  assignRef,
  createSetAdd,
  EMPTY_SET,
  createMapSet,
  EMPTY_MAP,
  createArrayAssign,
  createObjectIdentifierAssign,
  createObjectComputedAssign,
} from './context';
import isPrimitive from './is-primitive';
import isPromise from './is-promise';
import quote from './quote';
import serializePrimitive from './serialize-primitive';
import { AsyncServerValue } from './types';

export async function lookupRefsAsync(
  ctx: SerializationContext,
  current: AsyncServerValue,
) {
  if (isPrimitive(current)) {
    return;
  }
  if (isPromise(current)) {
    if (insertRef(ctx, current)) {
      await current.then((value) => lookupRefsAsync(ctx, value));
    }
  } else if (Array.isArray(current)) {
    if (insertRef(ctx, current)) {
      for (const value of current) {
        await lookupRefsAsync(ctx, value);
      }
    }
  } else if (constructorCheck<Set<AsyncServerValue>>(current, Set)) {
    if (insertRef(ctx, current)) {
      for (const item of current.keys()) {
        await lookupRefsAsync(ctx, item);
      }
    }
  } else if (constructorCheck<Map<AsyncServerValue, AsyncServerValue>>(current, Map)) {
    if (insertRef(ctx, current)) {
      for (const [key, value] of current.entries()) {
        await lookupRefsAsync(ctx, key);
        await lookupRefsAsync(ctx, value);
      }
    }
  } else if (
    (current.constructor === Object || current.constructor == null)
    && insertRef(ctx, current)
  ) {
    for (const value of Object.values(current)) {
      await lookupRefsAsync(ctx, value as AsyncServerValue);
    }
  }
}

export async function traverseAsync(
  ctx: SerializationContext,
  current: AsyncServerValue,
): Promise<string> {
  // Skip primitives
  if (isPrimitive(current)) {
    return serializePrimitive(current);
  }
  // Get reference of the current object
  let refResult = ctx.refs.get(current);
  if (refResult != null) {
    // Return the reference if already assigned a value
    if (ctx.assignedRefs.has(refResult)) {
      return getRefParam(ctx, refResult);
    }
    // Early sign (to prevent recursion)
    ctx.assignedRefs.add(refResult);
  }
  function getCurrentRef() {
    // This object received a ref, we must generate a setter ref
    if (refResult == null) {
      refResult = createRef(ctx, current);
    }
    return refResult;
  }
  function processRef(value: string) {
    if (refResult != null) {
      return assignRef(ctx, refResult, value);
    }
    return value;
  }
  if (isPromise(current)) {
    return current.then(async (value) => {
      ctx.stack.push(current);
      let result: string;
      if (ctx.stack.includes(value)) {
        const refParam = getRefParam(ctx, createRef(ctx, value));
        result = `Promise.resolve().then(()=>${refParam})`;
      } else {
        result = `Promise.resolve(${await traverseAsync(ctx, value)})`;
      }
      ctx.stack.pop();
      return processRef(result);
    });
  }
  // Transform Set
  if (constructorCheck<Set<AsyncServerValue>>(current, Set)) {
    const values: string[] = [];

    ctx.stack.push(current);
    for (const value of current.keys()) {
      if (ctx.stack.includes(value)) {
        // Received a ref, this might be a recursive ref, defer an assignment
        createSetAdd(ctx, getCurrentRef(), getRefParam(ctx, createRef(ctx, value)));
      } else {
        values.push(await traverseAsync(ctx, value));
      }
    }
    ctx.stack.pop();

    return processRef(values.length ? `new Set([${values.join(',')}])` : EMPTY_SET);
  }
  if (constructorCheck<Map<AsyncServerValue, AsyncServerValue>>(current, Map)) {
    const values: string[] = [];

    ctx.stack.push(current);
    for (const [key, value] of current.entries()) {
      if (ctx.stack.includes(key)) {
        const ref = getCurrentRef();
        const keyRef = getRefParam(ctx, createRef(ctx, key));
        if (ctx.stack.includes(value)) {
          const valueRef = getRefParam(ctx, createRef(ctx, value));
          createMapSet(ctx, ref, keyRef, valueRef);
        } else {
          const valueResult = await traverseAsync(ctx, value);
          createMapSet(ctx, ref, keyRef, valueResult);
        }
      } else if (ctx.stack.includes(value)) {
        const valueRef = getRefParam(ctx, createRef(ctx, value));
        const keyResult = await traverseAsync(ctx, key);
        createMapSet(ctx, getCurrentRef(), keyResult, valueRef);
      } else {
        const keyResult = await traverseAsync(ctx, key);
        const valueResult = await traverseAsync(ctx, value);
        values.push(`[${keyResult},${valueResult}]`);
      }
    }
    ctx.stack.pop();
    return processRef(values.length ? `new Map([${values.join(',')}])` : EMPTY_MAP);
  }
  if (Array.isArray(current)) {
    let values = '';

    ctx.stack.push(current);
    for (let i = 0, len = current.length; i < len; i += 1) {
      if (i in current) {
        const item = current[i];
        if (ctx.stack.includes(item)) {
          const refParam = getRefParam(ctx, createRef(ctx, item));
          createArrayAssign(ctx, getCurrentRef(), i, refParam);
          values += ',';
        } else {
          values += await traverseAsync(ctx, item);
          if (i < current.length - 1) {
            values += ',';
          }
        }
      } else {
        values += ',';
      }
    }
    ctx.stack.pop();
    return processRef(`[${values}]`);
  }
  const empty = current.constructor == null;
  if (current.constructor === Object || empty) {
    const values: string[] = [];

    ctx.stack.push(current);
    for (const [key, value] of Object.entries(current)) {
      const check = Number(key);
      // Test if key is a valid number or JS identifier
      // so that we don't have to serialize the key and wrap with brackets
      if (
        check >= 0
        || (Number.isNaN(check) && /^([$A-Z_][0-9A-Z_$]*)$/i.test(key))
      ) {
        if (ctx.stack.includes(value)) {
          const ref = getCurrentRef();
          const refParam = getRefParam(ctx, createRef(ctx, value));
          if (Number.isNaN(check)) {
            createObjectIdentifierAssign(ctx, ref, key, refParam);
          } else {
            createObjectComputedAssign(ctx, ref, key, refParam);
          }
        } else {
          const valueResult = await traverseAsync(ctx, value as AsyncServerValue);
          values.push(`${key}:${valueResult}`);
        }
      } else if (ctx.stack.includes(value)) {
        const refParam = getRefParam(ctx, createRef(ctx, value));
        createObjectComputedAssign(ctx, getCurrentRef(), quote(key), refParam);
      } else {
        const valueResult = await traverseAsync(ctx, value as AsyncServerValue);
        values.push(`${quote(key)}:${valueResult}`);
      }
    }
    ctx.stack.pop();
    const value = `{${values.join(',')}}`;
    if (empty) {
      return processRef(`Object.assign(Object.create(null),${value})`);
    }
    return processRef(value);
  }
  throw new Error('Unserializable value');
}
