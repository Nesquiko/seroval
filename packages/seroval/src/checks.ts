import { PrimitiveValue } from './types';

export function isPrimitive(current: unknown): current is PrimitiveValue {
  if (!current || current === true) {
    return true;
  }
  const type = typeof current;
  return type === 'number'
    || type === 'string'
    || type === 'bigint';
}

export function isPromise<T>(obj: unknown): obj is PromiseLike<T> {
  return !!obj
    && (typeof obj === 'object' || typeof obj === 'function')
    && 'then' in obj
    && typeof obj.then === 'function';
}

export function isIterable<T>(obj: unknown): obj is Iterable<T> {
  return !!obj
    && typeof obj === 'object'
    && Symbol.iterator in obj
    && typeof obj[Symbol.iterator] === 'function';
}
