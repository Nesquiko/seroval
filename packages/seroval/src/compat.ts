/* eslint-disable guard-for-in */
/**
 * References
 * - https://kangax.github.io/compat-table/es6/
 * - MDN
 * - ESBuild
 */

import assert from './assert';

export type Version = [major: number, minor: number, patch: number];

function compareVersion(left: Version, right: Version) {
  let diff = left[0] - right[0];
  if (diff === 0) {
    diff = left[1] - right[1];
  }
  if (diff === 0) {
    diff = left[2] - right[2];
  }
  return diff;
}

export interface PlatformVersion {
  es: Version;
  // desktop
  chrome?: Version;
  edge?: Version;
  safari?: Version;
  firefox?: Version;
  opera?: Version;
  // mobile
  ios?: Version;
  samsung?: Version;
  // js runtimes
  deno?: Version;
  node?: Version;
}

export type Platform = keyof PlatformVersion;

export const enum Feature {
  AggregateError = 0x01,
  ArrayPrototypeValues = 0x02,
  ArrowFunction = 0x04,
  BigInt = 0x08,
  Map = 0x10,
  MethodShorthand = 0x20,
  ObjectAssign = 0x40,
  Promise = 0x80,
  Set = 0x100,
  SymbolIterator = 0x200,
  TypedArray = 0x400,
  BigIntTypedArray = 0x800,
}

type VersionTable = {
  [key in Feature]: PlatformVersion;
}

const VERSION_TABLE: VersionTable = {
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AggregateError#browser_compatibility
  [Feature.AggregateError]: {
    es: [2021, 0, 0],
    chrome: [85, 0, 0],
    edge: [85, 0, 0],
    firefox: [79, 0, 0],
    opera: [71, 0, 0],
    safari: [14, 0, 0],
    ios: [14, 0, 0],
    samsung: [14, 0, 0],
    deno: [1, 2, 0],
    node: [15, 0, 0],
  },
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/values#browser_compatibility
  [Feature.ArrayPrototypeValues]: {
    es: [6, 0, 0],
    chrome: [66, 0, 0],
    edge: [14, 0, 0],
    firefox: [60, 0, 0],
    opera: [53, 0, 0],
    safari: [9, 0, 0],
    ios: [9, 0, 0],
    samsung: [9, 0, 0],
    deno: [1, 0, 0],
    node: [10, 9, 0],
  },
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions#browser_compatibility
  [Feature.ArrowFunction]: {
    es: [6, 0, 0],
    chrome: [45, 0, 0],
    edge: [12, 0, 0],
    firefox: [24, 0, 0],
    opera: [32, 0, 0],
    safari: [10, 0, 0],
    ios: [10, 0, 0],
    samsung: [5, 0, 0],
    deno: [1, 0, 0],
    node: [4, 0, 0],
  },
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt#browser_compatibility
  [Feature.BigInt]: {
    es: [2020, 0, 0],
    chrome: [67, 0, 0],
    edge: [79, 0, 0],
    firefox: [68, 0, 0],
    opera: [54, 0, 0],
    safari: [14, 0, 0],
    ios: [14, 0, 0],
    samsung: [9, 0, 0],
    deno: [1, 0, 0],
    node: [10, 4, 0],
  },
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map#browser_compatibility
  [Feature.Map]: {
    es: [6, 0, 0],
    chrome: [38, 0, 0],
    edge: [12, 0, 0],
    firefox: [13, 0, 0],
    opera: [25, 0, 0],
    safari: [8, 0, 0],
    ios: [8, 0, 0],
    samsung: [3, 0, 0],
    deno: [1, 0, 0],
    node: [0, 12, 0],
  },
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Method_definitions#browser_compatibility
  [Feature.MethodShorthand]: {
    es: [6, 0, 0],
    chrome: [39, 0, 0],
    edge: [12, 0, 0],
    firefox: [34, 0, 0],
    opera: [26, 0, 0],
    safari: [9, 0, 0],
    ios: [9, 0, 0],
    samsung: [4, 0, 0],
    deno: [1, 0, 0],
    node: [4, 0, 0],
  },
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign#browser_compatibility
  [Feature.ObjectAssign]: {
    es: [6, 0, 0],
    chrome: [45, 0, 0],
    edge: [12, 0, 0],
    firefox: [34, 0, 0],
    opera: [32, 0, 0],
    safari: [9, 0, 0],
    ios: [9, 0, 0],
    samsung: [5, 0, 0],
    deno: [1, 0, 0],
    node: [4, 0, 0],
  },
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#browser_compatibility
  [Feature.Promise]: {
    es: [6, 0, 0],
    chrome: [32, 0, 0],
    edge: [12, 0, 0],
    firefox: [29, 0, 0],
    opera: [19, 0, 0],
    safari: [8, 0, 0],
    ios: [8, 0, 0],
    samsung: [2, 0, 0],
    deno: [1, 0, 0],
    node: [0, 12, 0],
  },
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set#browser_compatibility
  [Feature.Set]: {
    es: [6, 0, 0],
    chrome: [38, 0, 0],
    edge: [12, 0, 0],
    firefox: [13, 0, 0],
    opera: [25, 0, 0],
    safari: [8, 0, 0],
    ios: [8, 0, 0],
    samsung: [3, 0, 0],
    deno: [1, 0, 0],
    node: [0, 12, 0],
  },
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/iterator#browser_compatibility
  [Feature.SymbolIterator]: {
    es: [6, 0, 0],
    chrome: [43, 0, 0],
    edge: [12, 0, 0],
    firefox: [36, 0, 0],
    opera: [30, 0, 0],
    safari: [10, 0, 0],
    ios: [10, 0, 0],
    samsung: [4, 0, 0],
    deno: [1, 0, 0],
    node: [0, 12, 0],
  },
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray#browser_compatibility
  [Feature.TypedArray]: {
    es: [6, 0, 0],
    chrome: [7, 0, 0],
    edge: [12, 0, 0],
    firefox: [4, 0, 0],
    opera: [11, 6, 0],
    safari: [5, 1, 0],
    ios: [4, 2, 0],
    samsung: [1, 0, 0],
    deno: [1, 0, 0],
    node: [0, 10, 0],
  },
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt64Array#browser_compatibility
  [Feature.BigIntTypedArray]: {
    es: [2020, 0, 0],
    chrome: [67, 0, 0],
    edge: [79, 0, 0],
    firefox: [68, 0, 0],
    opera: [54, 6, 0],
    safari: [15, 1, 0],
    ios: [15, 2, 0],
    samsung: [9, 0, 0],
    deno: [1, 0, 0],
    node: [10, 4, 0],
  },
};

const TARGET_REGEX = /^(es|chrome|edge|safari|firefox|opera|ios|samsung|deno|node)([0-9]+(\.[0-9]+(\.[0-9]+)?)?)$/i;

export type Target = [platform: Platform, version: Version];

function parseTarget(target: string): Target {
  const result = TARGET_REGEX.exec(target);
  assert(result, `Invalid target "${target}"`);
  const [, platform, version] = result;
  const [major, minor = '0', patch = '0'] = version.split('.');
  return [platform as Platform, [Number(major), Number(minor), Number(patch)]];
}

function getTargetVersions(targets: string | string[]): Target[] {
  if (Array.isArray(targets)) {
    const versions: Target[] = [];
    for (const target of targets) {
      versions.push(parseTarget(target));
    }
    return versions;
  }
  return [parseTarget(targets)];
}

export function parseTargets(targets: string | string[]): number {
  const parsed = getTargetVersions(targets);
  let flags = 0;

  for (const key in VERSION_TABLE) {
    const base = VERSION_TABLE[key as unknown as Feature];
    let flag = true;
    for (const [platform, version] of parsed) {
      const baseVersion = base[platform];
      flag = flag && !!baseVersion && (compareVersion(baseVersion, version) <= 0);
    }
    if (flag) {
      flags |= Number(key);
    }
  }

  return flags;
}
