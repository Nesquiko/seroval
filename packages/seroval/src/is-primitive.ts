import { PrimitiveValue } from './types';

export default function isPrimitive(current: unknown): current is PrimitiveValue {
  if (!current || current === true) {
    return true;
  }
  const type = typeof current;
  const classConst = current.constructor;
  return type === 'number'
    || type === 'string'
    || type === 'bigint'
    || classConst === Date;
}
