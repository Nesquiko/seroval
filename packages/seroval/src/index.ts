import { forEach, join } from './array';
import getIdentifier from './get-identifier';
import quote from './quote';

const EMPTY_SET = 'new Set';
const EMPTY_MAP = 'new Map';
const EMPTY_ARRAY = '[]';

// Values that are non-recursive
export type PrimitiveValue =
  | boolean
  | string
  | number
  | undefined
  | null
  | bigint
  | Date
  | RegExp;

export type ServerValue =
  | PrimitiveValue
  | Array<ServerValue>
  | { [key: string | number]: ServerValue }
  | Set<ServerValue>
  | Map<ServerValue, ServerValue>;

// Reference counter
const REF_COUNT = new Map<ServerValue, number>();

const STACK = new Set();

// Value-to-ref map
const REFS = new Map<ServerValue, number>();

// Set to check if ref id already has an assigned value
const ASSIGNED_REFS = new Set<number>();

// Variables
let VARS: string[] = [];

// Array of assignments to be done (used for recursion)
type Assignment = [source: string, value: string];
let ASSIGNMENTS: Assignment[] = [];

// Array of Map.prototype.set calls
type MapSet = [source: string, key: string, value: string];
let MAP_SETS: MapSet[] = [];

// Array of Set.prototype.add calls
type SetAdd = [source: string, value: string];
let SET_ADDS: SetAdd[] = [];

function reset() {
  ASSIGNMENTS = [];
  MAP_SETS = [];
  SET_ADDS = [];
  VARS = [];
  REF_COUNT.clear();
  REFS.clear();
  ASSIGNED_REFS.clear();
}

function resolveAssignments() {
  if (ASSIGNMENTS.length) {
    const result: Record<string, string> = {};

    // Merge all assignments with similar source
    forEach(ASSIGNMENTS, ([source, value]) => {
      if (value in result) {
        result[value] = `${source}=${result[value]}`;
      } else {
        result[value] = `${source}=${value}`;
      }
    });

    return `${join(Object.values(result), ',')},`;
  }
  return '';
}

function resolveMapSets() {
  if (MAP_SETS.length) {
    const result: Record<string, string> = {};

    // Merge all assignments with similar source
    forEach(MAP_SETS, ([source, key, value]) => {
      if (source in result) {
        result[source] = `${result[source]}.set(${key},${value})`;
      } else {
        result[source] = `${source}.set(${key},${value})`;
      }
    });

    return `${join(Object.values(result), ',')},`;
  }
  return '';
}

function resolveSetAdds() {
  if (SET_ADDS.length) {
    const result: Record<string, string> = {};

    // Merge all assignments with similar source
    forEach(SET_ADDS, ([source, value]) => {
      if (source in result) {
        result[source] = `${result[source]}.add(${value})`;
      } else {
        result[source] = `${source}.add(${value})`;
      }
    });

    return `${join(Object.values(result), ',')},`;
  }
  return '';
}

function resolvePatches() {
  return `${resolveAssignments()}${resolveMapSets()}${resolveSetAdds()}`;
}

function insertRef(current: ServerValue) {
  const count = REF_COUNT.get(current) || 0;
  REF_COUNT.set(current, count + 1);
  return count === 0;
}

function constructorCheck<T extends NonNullable<ServerValue>>(
  value: NonNullable<ServerValue>,
  constructor: unknown,
): value is T {
  return value.constructor === constructor;
}

function isPrimitive(current: ServerValue): current is PrimitiveValue {
  if (!current || current === true) {
    return true;
  }
  const type = typeof current;
  const classConst = current.constructor;
  return type === 'number'
    || type === 'string'
    || type === 'bigint'
    || classConst === Date
    || classConst === RegExp;
}

function lookupRefs(current: ServerValue) {
  if (isPrimitive(current)) {
    return;
  }
  if (Array.isArray(current)) {
    if (insertRef(current)) {
      forEach(current, lookupRefs);
    }
  } else if (constructorCheck<Set<ServerValue>>(current, Set)) {
    if (insertRef(current)) {
      for (const item of current.keys()) {
        lookupRefs(item);
      }
    }
  } else if (constructorCheck<Map<ServerValue, ServerValue>>(current, Map)) {
    if (insertRef(current)) {
      for (const [key, value] of current.entries()) {
        lookupRefs(key);
        lookupRefs(value);
      }
    }
  } else if (current.constructor === Object && insertRef(current)) {
    forEach(Object.values(current), lookupRefs);
  }
}

function createRef(value: ServerValue) {
  const current = REFS.get(value);
  if (current != null) {
    return current;
  }
  const id = REFS.size;
  REFS.set(value, id);
  return id;
}

function getRefParam(index: number) {
  if (VARS[index]) {
    return VARS[index];
  }
  const result = getIdentifier(index);
  VARS[index] = result;
  return result;
}

function assignRef(index: number, value: string) {
  return `${getRefParam(index)}=${value}`;
}

function createAssignment(source: string, value: string) {
  ASSIGNMENTS.push([source, value]);
}

function createSetAdd(ref: number, value: string) {
  SET_ADDS.push([getRefParam(ref), value]);
}

function createMapSet(ref: number, key: string, value: string) {
  MAP_SETS.push([getRefParam(ref), key, value]);
}

function createArrayAssign(ref: number, index: number, value: string) {
  createAssignment(`${getRefParam(ref)}[${index}]`, value);
}

function createObjectIdentifierAssign(ref: number, key: string, value: string) {
  createAssignment(`${getRefParam(ref)}.${key}`, value);
}

function createObjectStringAssign(ref: number, key: string, value: string) {
  createAssignment(`${getRefParam(ref)}[${quote(key)}]`, value);
}

function serializePrimitive(value: PrimitiveValue): string {
  // Shortened forms
  if (value === true) {
    return '!0';
  }
  if (value === false) {
    return '!1';
  }
  if (value === undefined) {
    return 'void 0';
  }
  if (value === null) {
    return 'null';
  }
  // negative 0 isn't the same as 0
  if (Object.is(value, -0)) {
    return '-0';
  }
  if (typeof value === 'bigint') {
    return `BigInt("${value}")`;
  }
  if (typeof value === 'string') {
    return quote(value);
  }
  if (value.constructor === Date) {
    return `new Date("${value.toISOString()}")`;
  }
  return String(value);
}

function traverse(current: ServerValue): string | { referred: string } {
  // Skip primitives
  if (isPrimitive(current)) {
    return serializePrimitive(current);
  }
  // Get reference of the current object
  let refResult = REFS.get(current);
  if (refResult != null) {
    // Return the reference if already assigned a value
    if (ASSIGNED_REFS.has(refResult)) {
      return { referred: getRefParam(refResult) };
    }
    // Early sign (to prevent recursion)
    ASSIGNED_REFS.add(refResult);
  }
  // Transform Set
  if (constructorCheck<Set<ServerValue>>(current, Set)) {
    if (current.size) {
      const values: string[] = [];

      STACK.add(current);
      for (const value of current.keys()) {
        if (STACK.has(value)) {
          // This object received a ref, we must generate a setter ref
          if (refResult == null) {
            refResult = createRef(current);
          }
          // Received a ref, this might be a recursive ref, defer an assignment
          createSetAdd(refResult, getRefParam(createRef(value)));
        } else {
          const result = traverse(value);
          if (typeof result === 'string') {
            values.push(result);
          } else {
            values.push(result.referred);
          }
        }
      }
      STACK.delete(current);

      const value = values.length ? `new Set([${join(values, ',')}])` : EMPTY_SET;
      if (refResult != null) {
        return assignRef(refResult, value);
      }
      return value;
    }
    return EMPTY_SET;
  }
  if (constructorCheck<Map<ServerValue, ServerValue>>(current, Map)) {
    if (current.size) {
      const values: string[] = [];

      STACK.add(current);
      for (const [key, value] of current.entries()) {
        if (STACK.has(key)) {
          if (refResult == null) {
            refResult = createRef(current);
          }
          const keyRef = getRefParam(createRef(key));
          if (STACK.has(value)) {
            createMapSet(refResult, keyRef, getRefParam(createRef(value)));
          } else {
            const valueResult = traverse(value);
            if (typeof valueResult === 'string') {
              createMapSet(refResult, keyRef, valueResult);
            } else {
              createMapSet(refResult, keyRef, valueResult.referred);
            }
          }
        } else if (STACK.has(value)) {
          if (refResult == null) {
            refResult = createRef(current);
          }
          const valueRef = getRefParam(createRef(value));
          const keyResult = traverse(key);
          if (typeof keyResult === 'string') {
            createMapSet(refResult, keyResult, valueRef);
          } else {
            createMapSet(refResult, keyResult.referred, valueRef);
          }
        } else {
          const keyResult = traverse(key);
          const valueResult = traverse(value);
          if (typeof keyResult === 'string') {
            if (typeof valueResult === 'string') {
              values.push(`[${keyResult},${valueResult}]`);
            } else {
              values.push(`[${keyResult},${valueResult.referred}]`);
            }
          } else {
            if (typeof valueResult === 'string') {
              values.push(`[${keyResult.referred},${valueResult}]`);
            } else {
              values.push(`[${keyResult.referred},${valueResult.referred}]`);
            }
          }
        }
      }
      STACK.delete(current);
      const value = values.length ? `new Map([${join(values, ',')}])` : EMPTY_MAP;
      if (refResult != null) {
        return assignRef(refResult, value);
      }
      return value;
    }
    return EMPTY_MAP;
  }
  if (Array.isArray(current)) {
    if (current.length) {
      let values = '';

      STACK.add(current);
      forEach(current, (item, i) => {
        if (i in current) {
          if (STACK.has(item)) {
            if (refResult == null) {
              refResult = createRef(current);
            }
            createArrayAssign(refResult, i, getRefParam(createRef(item)));
            values += ',';
          } else {
            const result = traverse(item);
            if (typeof result === 'string') {
              values += result;
            } else {
              values += result.referred;
            }
            if (i < current.length - 1) {
              values += ',';
            }
          }
        } else {
          values += ',';
        }
      });
      STACK.delete(current);

      const value = `[${values}]`;
      if (refResult != null) {
        return assignRef(refResult, value);
      }
      return value;
    }
    return EMPTY_ARRAY;
  }
  if (current.constructor === Object) {
    const values: string[] = [];

    STACK.add(current);
    forEach(Object.entries(current), ([key, value]) => {
      const check = Number(key);
      // Test if key is a valid number or JS identifier
      // so that we don't have to serialize the key and wrap with brackets
      if (Number.isNaN(check) || /^([$A-Z_][0-9A-Z_$]*)$/i.test(key)) {
        if (STACK.has(value)) {
          if (refResult == null) {
            refResult = createRef(current);
          }
          createObjectIdentifierAssign(refResult, key, getRefParam(createRef(value)));
        } else {
          const result = traverse(value);
          if (typeof result === 'string') {
            values.push(`${key}:${result}`);
          } else {
            values.push(`${key}:${result.referred}`);
          }
        }
      } else if (STACK.has(value)) {
        if (refResult == null) {
          refResult = createRef(current);
        }
        createObjectStringAssign(refResult, key, getRefParam(createRef(value)));
      } else {
        const result = traverse(value);
        if (typeof result === 'string') {
          values.push(`${quote(key)}:${result}`);
        } else {
          values.push(`${quote(key)}:${result.referred}`);
        }
      }
    });
    STACK.delete(current);

    const value = `{${join(values, ',')}}`;
    if (refResult != null) {
      return assignRef(refResult, value);
    }
    return value;
  }
  throw new Error('Unserializable value');
}

export default function serialize(source: ServerValue) {
  if (isPrimitive(source)) {
    return serializePrimitive(source);
  }
  reset();
  // Lookup possible shared references
  lookupRefs(source);
  // Add shared references to refs
  for (const [key, value] of REF_COUNT.entries()) {
    if (value > 1) {
      createRef(key);
    }
  }
  // Get top-level serialization
  const result = traverse(source);
  if (typeof result !== 'string') {
    throw new Error('Unreachable error');
  }

  // Shared references detected
  if (VARS.length) {
    // Get (or create) a ref from the source
    const index = getRefParam(createRef(source));
    const patches = resolvePatches();
    const params = VARS.length > 1
      ? `(${join(VARS, ',')})`
      : VARS[0];
    // Source is probably already assigned
    if (result.startsWith(index)) {
      return `(${params}=>(${result},${patches}${index}))()`;
    }
    return `(${params}=>(${index}=${result},${patches}${index}))()`;
  }
  if (source.constructor === Object) {
    return `(${result})`;
  }
  return result;
}
