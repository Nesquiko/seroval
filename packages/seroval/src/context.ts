import { ALL_ENABLED } from './compat';
import getIdentifier from './get-identifier';
import { AsyncServerValue } from './types';

interface IndexAssignment {
  t: 'index';
  s: string;
  k: undefined;
  v: string;
}

interface MapAssignment {
  t: 'map';
  s: string;
  k: string;
  v: string;
}

interface SetAssignment {
  t: 'set';
  s: string;
  k: undefined;
  v: string;
}

// Array of assignments to be done (used for recursion)
export type Assignment =
  | IndexAssignment
  | MapAssignment
  | SetAssignment;

export interface ParserContext {
  refs: Map<unknown, number>;
  markedRefs: Set<number>;
  features: number;
}

export interface SerializationContext {
  stack: number[];
  // Map tree refs to actual refs
  validRefs: number[];
  refSize: number;
  // Refs that are...referenced
  markedRefs: Set<number>;
  // Variables
  vars: string[];
  // Array of assignments to be done (used for recursion)
  assignments: Assignment[];
  // Supported features
  features: number;

  valueMap: Map<number, AsyncServerValue>;
}

export interface Options {
  disabledFeatures: number;
}

const DEFAULT_OPTIONS: Options = {
  disabledFeatures: 0,
};

export function createParserContext(options: Partial<Options> = {}): ParserContext {
  // eslint-disable-next-line prefer-object-spread
  const result = Object.assign({}, DEFAULT_OPTIONS, options || {});
  return {
    markedRefs: new Set(),
    refs: new Map(),
    features: ALL_ENABLED ^ result.disabledFeatures,
  };
}

export interface SerializationOptions {
  markedRefs: number[] | Set<number>;
  features: number;
}

export function createSerializationContext(options: SerializationOptions): SerializationContext {
  return {
    stack: [],
    vars: [],
    assignments: [],
    validRefs: [],
    refSize: 0,
    features: options.features,
    markedRefs: new Set(options.markedRefs),
    valueMap: new Map(),
  };
}

/**
 * Increments the number of references the referenced value has
 */
export function markRef(ctx: ParserContext | SerializationContext, current: number) {
  ctx.markedRefs.add(current);
}

/**
 * Creates a new reference ID from a given reference ID
 * This new reference ID means that the reference itself
 * has been referenced at least once, and is used to generate
 * the variables
 */
function createValidRef(
  ctx: SerializationContext,
  index: number,
) {
  const current = ctx.validRefs[index];
  if (current == null) {
    const value = ctx.refSize++;
    ctx.validRefs[index] = value;
    return value;
  }
  return current;
}

/**
 * Creates the reference param (identifier) from the given reference ID
 * Calling this function means the value has been referenced somewhere
 */
export function getRefParam(ctx: SerializationContext, index: number) {
  const actualIndex = createValidRef(ctx, index);
  if (ctx.vars[actualIndex]) {
    return ctx.vars[actualIndex];
  }
  const result = getIdentifier(actualIndex);
  ctx.vars[actualIndex] = result;
  return result;
}

/**
 * Creates a reference ID from the given values
 */
export function createRef(
  ctx: ParserContext,
  current: unknown,
  mark: boolean,
): number {
  // Check if reference number already exists
  const ref = ctx.refs.get(current);
  if (ref != null) {
    // Exists, means this value is currently
    // being referenced
    // Mark reference
    if (mark) {
      markRef(ctx, ref);
    }
    return ref;
  }
  // Create a new reference ID
  const id = ctx.refs.size;
  ctx.refs.set(current, id);
  return id;
}
