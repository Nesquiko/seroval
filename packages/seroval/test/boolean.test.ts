import { describe, it, expect } from 'vitest';
import { serialize, serializeAsync } from '../src';

describe('boolean', () => {
  describe('serialize', () => {
    it('supports boolean', () => {
      expect(serialize(true)).toBe('!0');
      expect(serialize(false)).toBe('!1');
    });
  });
  describe('serializeAsync', () => {
    it('supports boolean', async () => {
      expect(await serializeAsync(true)).toBe('!0');
      expect(await serializeAsync(false)).toBe('!1');
    });
  });
});
