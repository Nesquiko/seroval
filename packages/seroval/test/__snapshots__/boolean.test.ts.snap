// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`boolean > crossSerialize > supports boolean 1`] = `"!0"`;

exports[`boolean > crossSerialize > supports boolean 2`] = `"!1"`;

exports[`boolean > crossSerializeAsync > supports boolean 1`] = `"$R[0]=Promise.resolve(!0)"`;

exports[`boolean > crossSerializeAsync > supports boolean 2`] = `"$R[0]=Promise.resolve(!1)"`;

exports[`boolean > crossSerializeAsync > supports false value 1`] = `"$R[0]=$P()"`;

exports[`boolean > crossSerializeAsync > supports false value 2`] = `"$Ps(0,!1)"`;

exports[`boolean > crossSerializeAsync > supports true value 1`] = `"$R[0]=$P()"`;

exports[`boolean > crossSerializeAsync > supports true value 2`] = `"$Ps(0,!0)"`;
